pkg/admissioncontroller/webhook/admission/internaldomainsecret/handler.go

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internaldomainsecret
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	gardercore "github.com/gardener/gardener/pkg/apis/core"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	gutil "github.com/gardener/gardener/pkg/utils/gardener"
+)
+
+// Handler validates the immutability of the internal domain secret.
+type Handler struct {
+	Logger    logr.Logger
+	APIReader client.Reader
+}
+
+// ValidateCreate performs the check.
+func (h *Handler) ValidateCreate(ctx context.Context, obj runtime.Object) error {
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+
+	secret, ok := obj.(*corev1.Secret)
+	if !ok {
+		return fmt.Errorf("expected *corev1.Secret but got %T", obj)
+	}
+
+	alreadyExists, err := h.secretAlreadyExists(ctx, secret)
+	if err != nil {
+		return err
+	}
+	if alreadyExists {
+		return nil
+	}
```

Comment:
Do we still need this check if down below in `internalDomainSecretExists` the search is restricted to the secret's namespace?

Reply:
Probably not, I remove it.